lib/Pipes/PipelineManager.cpp

```diff
@@ -132,46 +125,40 @@ PipelineManager::create(llvm::ArrayRef<std::string> Pipelines,
       return MaybeBuffer.takeError();
     LoadedPipelines.emplace_back((*MaybeBuffer)->getBuffer().str());
   }
-  if (auto MaybePipeline = setUpPipeline(*Manager.PipelineContext,
-                                         *Manager.Loader,
-                                         LoadedPipelines,
-                                         ExecutionDirectory);
-      MaybePipeline)
-    Manager.Runner = make_unique<pipeline::Runner>(std::move(*MaybePipeline));
-  else
-    return MaybePipeline.takeError();
 
-  Manager.recalculateAllPossibleTargets();
-  return std::move(Manager);
+  return createFromMemory(LoadedPipelines, EnablingFlags, ExecutionDirectory);
 }
 
-llvm::Expected<PipelineManager>
-PipelineManager::createContexts(llvm::ArrayRef<std::string> EnablingFlags,
-                                llvm::StringRef ExecutionDirectory) {
-  PipelineManager Manager;
-  Manager.ExecutionDirectory = ExecutionDirectory.str();
-  Manager.Context = std::make_unique<llvm::LLVMContext>();
-  auto Ctx = setUpContext(*Manager.Context);
-  Manager.PipelineContext = make_unique<pipeline::Context>(std::move(Ctx));
-
-  auto Loader = setupLoader(*Manager.PipelineContext, EnablingFlags);
-  Manager.Loader = make_unique<pipeline::Loader>(std::move(Loader));
-  return Manager;
+PipelineManager::PipelineManager(llvm::ArrayRef<std::string> EnablingFlags,
+                                 std::unique_ptr<revng::StorageClient>
+                                   &&Client) :
+  StorageClient(std::move(Client)),
+  ExecutionDirectory(StorageClient.get(), "") {
+  Context = std::make_unique<llvm::LLVMContext>();
+  auto Ctx = setUpContext(*Context);
+  PipelineContext = make_unique<pipeline::Context>(std::move(Ctx));
+
+  auto Loader = setupLoader(*PipelineContext, EnablingFlags);
+  this->Loader = make_unique<pipeline::Loader>(std::move(Loader));
 }
 
 llvm::Expected<PipelineManager>
 PipelineManager::createFromMemory(llvm::ArrayRef<std::string> PipelineContent,
                                   llvm::ArrayRef<std::string> EnablingFlags,
                                   llvm::StringRef ExecutionDirectory) {
+  std::unique_ptr<revng::StorageClient> Client;
+  if (not ExecutionDirectory.empty()) {
+    auto MaybeClient = revng::StorageClient::fromURL(ExecutionDirectory);
+    if (!MaybeClient)
+      return MaybeClient.takeError();
+    Client = std::move(MaybeClient.get());
+  }
```

Comment:
PipelineManager is the god object of the entire pipeline, there's nothing above. Also this needs to somehow work with python. Moving it besides like `rp_make_storage_client` to then just pass it to the manager seems a bit redundant IMHO

Reply:
> PipelineManager is the god object of the entire pipeline, there's nothing above

There are `main`s, which is the appropriate place where these things need to be handled.

> rp_make_storage_client

The C API can do it. I'm just suggesting to enforce the invariant that there no paths/URLs anywhere from `PipelineManager` below, just `revng::Path`.

Reply:
> just revng::Path

But `PipelineManager` has to `StorageClient::commit` (unless mains suddenly decide, but I think is poor coupling), I think having this level of indirection is unnecessary

Reply:
I'm suggesting to:
1. Have `PipelineManager` get a `std::move`'d `StorageClient` in the constructor for the execution directory. Then, it can call `commit` whenever appropriate.
2. Any other method that does not implicitly use the execution directory, accept a `revng::Path`.

Reply:
Doing (1) just means duplicating a lot of code around to call `StorageClient::fromURL`, I think it worsens the API
For (2) that's already happening, the only places where there's `StringRef` for paths is `PipelineManager::create*` and `Runner::storeToDiskDebug`

Reply:
Strings should not be the only way to instantiate `PipelineManager`, otherwise we're coupling `PipelineManager` to the currently known list of `StorageClient`s. One day in testing we might have some custom `StorageClient`.

You can either make the constructor public or add another factory method that takes `StorageClient &&`.

Maybe `storeToDiskDebug` should be rename to `dump`.

Reply:
Added an overload to `createFromMemory` that accepts a `StorageClient&&`
Renamed `storeToDiskDebug` to `dump`